File: packages/typescript-sdk/src/stream.ts
/**
 * StreamClient — real-time balance and transaction streaming via WebSocket/SSE.
 * 
 * Uses WebSocket as the primary transport with SSE as a fallback.
 * Provides live balance updates, transaction notifications, and agent status changes.
 */
import { MogBankClient } from './client';
import { WSMessage, WSMessageType } from './types';

export type StreamEventHandler = (data: unknown) => void;

export interface StreamSubscription {
  channel: string;
  handler: StreamEventHandler;
  unsubscribe: () => void;
}

export class StreamClient {
  private ws: WebSocket | null = null;
  private eventSource: EventSource | null = null;
  private subscriptions: Map<string, Set<StreamEventHandler>> = new Map();
  private reconnectAttempts = 0;
  private maxReconnectAttempts = 10;
  private reconnectDelay = 1000;
  private isConnected = false;
  private pingInterval: ReturnType<typeof setInterval> | null = null;

  constructor(private client: MogBankClient) {}

  /**
   * Connect to the balance streaming endpoint.
   * Tries WebSocket first, falls back to SSE.
   */
  async connect(): Promise<void> {
    const wsUrl = this.client.config.baseUrl
      .replace(/^http/, 'ws')
      .replace(/\/$/, '');

    try {
      await this.connectWebSocket(`${wsUrl}/stream/ws`);
    } catch {
      console.warn('[MogBank SDK] WebSocket unavailable, falling back to SSE');
      this.connectSSE(`${this.client.config.baseUrl}/stream/sse`);
    }
  }

  /**
   * Connect via WebSocket with authentication.
   */
  private connectWebSocket(url: string): Promise<void> {
    return new Promise((resolve, reject) => {
      try {
        // Add agent ID as query param for authentication
        const authUrl = this.client.config.agentId
          ? `${url}?agent_id=${encodeURIComponent(this.client.config.agentId)}`
          : url;

        this.ws = new WebSocket(authUrl);

        this.ws.onopen = () => {
          this.isConnected = true;
          this.reconnectAttempts = 0;
          this.reconnectDelay = 1000;
          this.resubscribeAll();
          this.startPing();
          resolve();
        };

        this.ws.onmessage = (event: MessageEvent) => {
          try {
            const message: WSMessage = JSON.parse(event.data);
            this.handleMessage(message);
          } catch {
            console.warn('[MogBank SDK] Failed to parse WebSocket message');
          }
        };

        this.ws.onerror = () => {
          this.isConnected = false;
          reject(new Error('WebSocket connection failed'));
        };

        this.ws.onclose = () => {
          this.isConnected = false;
          this.stopPing();
          this.attemptReconnect(() => this.connectWebSocket(url));
        };
      } catch (error) {
        reject(error);
      }
    });
  }

  /**
   * Connect via SSE (Server-Sent Events) as a fallback.
   */
  private connectSSE(url: string): void {
    const authUrl = this.client.config.agentId
      ? `${url}?agent_id=${encodeURIComponent(this.client.config.agentId)}`
      : url;

    this.eventSource = new EventSource(authUrl);

    this.eventSource.onopen = () => {
      this.isConnected = true;
      this.reconnectAttempts = 0;
      this.reconnectDelay = 1000;
      this.resubscribeAll();
    };

    // SSE channels map to custom event types
    const channels = ['balance_update', 'transaction_update', 'agent_update'];
    for (const channel of channels) {
      this.eventSource.addEventListener(channel, (event: MessageEvent) => {
        try {
          const data = JSON.parse(event.data);
          this.dispatchToHandlers(channel, data);
        } catch {
          console.warn(`[MogBank SDK] Failed to parse SSE message for ${channel}`);
        }
      });
    }

    this.eventSource.onerror = () => {
      this.isConnected = false;
      this.attemptReconnect(() => this.connectSSE(url));
    };
  }

  /**
   * Handle incoming WebSocket messages and dispatch to subscribers.
   */
  private handleMessage(message: WSMessage): void {
    if (message.type === 'heartbeat') {
      return; // Silently acknowledge heartbeats
    }

    if (message.channel) {
      this.dispatchToHandlers(message.channel, message.data);
    }

    // Also dispatch to generic message type handler
    this.dispatchToHandlers(message.type, message.data);
  }

  /**
   * Dispatch data to all handlers subscribed to a channel.
   */
  private dispatchToHandlers(channel: string, data: unknown): void {
    const handlers = this.subscriptions.get(channel);
    if (handlers) {
      for (const handler of handlers) {
        try {
          handler(data);
        } catch (error) {
          console.error(`[MogBank SDK] Error in stream handler for ${channel}:`, error);
        }
      }
    }
  }

  /**
   * Subscribe to real-time updates for a channel.
   * 
   * Available channels:
   * - `balance_update` — fired when the agent's wallet balance changes
   * - `transaction_update` — fired when a transaction status changes
   * - `agent_update` — fired when agent KYA score or status changes
   * 
   * Returns an object with an `unsubscribe()` method.
   */
  subscribe(channel: WSMessageType | string, handler: StreamEventHandler): StreamSubscription {
    if (!this.subscriptions.has(channel)) {
      this.subscriptions.set(channel, new Set());
    }
    this.subscriptions.get(channel)!.add(handler);

    // If connected, send subscribe message over WebSocket
    if (this.isConnected && this.ws && this.ws.readyState === WebSocket.OPEN) {
      this.ws.send(JSON.stringify({
        type: 'subscribe',
        channel,
      }));
    }

    return {
      channel,
      handler,
      unsubscribe: () => this.unsubscribe(channel, handler),
    };
  }

  /**
   * Unsubscribe a specific handler from a channel.
   */
  unsubscribe(channel: string, handler: StreamEventHandler): void {
    const handlers = this.subscriptions.get(channel);
    if (handlers) {
      handlers.delete(handler);
      if (handlers.size === 0) {
        this.subscriptions.delete(channel);
      }
    }

    // If connected and no more handlers for this channel, send unsubscribe
    if (
      this.isConnected &&
      this.ws &&
      this.ws.readyState === WebSocket.OPEN &&
      !this.subscriptions.has(channel)
    ) {
      this.ws.send(JSON.stringify({
        type: 'unsubscribe',
        channel,
      }));
    }
  }

  /**
   * Subscribe to balance updates for a specific wallet.
   */
  onBalanceUpdate(handler: (data: { wallet_id: string; balance: number; currency: string }) => void): StreamSubscription {
    return this.subscribe('balance_update', handler as StreamEventHandler);
  }

  /**
   * Subscribe to transaction updates for the authenticated agent.
   */
  onTransactionUpdate(handler: (data: {
    transaction_id: string;
    status: string;
    tx_hash?: string;
    settled_at?: string;
  }) => void): StreamSubscription {
    return this.subscribe('transaction_update', handler as StreamEventHandler);
  }

  /**
   * Resubscribe all channels after reconnect.
   */
  private resubscribeAll(): void {
    for (const channel of this.subscriptions.keys()) {
      if (this.ws && this.ws.readyState === WebSocket.OPEN) {
        this.ws.send(JSON.stringify({
          type: 'subscribe',
          channel,
        }));
      }
    }
  }

  /**
   * Start sending ping/heartbeat messages to keep the connection alive.
   */
  private startPing(): void {
    this.pingInterval = setInterval(() => {
      if (this.ws && this.ws.readyState === WebSocket.OPEN) {
        this.ws.send(JSON.stringify({ type: 'heartbeat', timestamp: new Date().toISOString() }));
      }
    }, 30000);
  }

  /**
   * Stop the ping interval.
   */
  private stopPing(): void {
    if (this.pingInterval) {
      clearInterval(this.pingInterval);
      this.pingInterval = null;
    }
  }

  /**
   * Attempt to reconnect with exponential backoff.
   */
  private attemptReconnect(connectFn: () => void | Promise<void>): void {
    if (this.reconnectAttempts >= this.maxReconnectAttempts) {
      console.error('[MogBank SDK] Max reconnect attempts reached');
      return;
    }

    this.reconnectAttempts++;
    const delay = this.reconnectDelay * Math.pow(1.5, this.reconnectAttempts - 1);
    console.log(`[MogBank SDK] Reconnecting in ${delay}ms (attempt ${this.reconnectAttempts})`);

    setTimeout(() => {
      connectFn();
    }, delay);
  }

  /**
   * Disconnect from the streaming endpoint and clean up.
   */
  disconnect(): void {
    this.stopPing();

    if (this.ws) {
      this.ws.close(1000, 'Client disconnect');
      this.ws = null;
    }

    if (this.eventSource) {
      this.eventSource.close();
      this.eventSource = null;
    }

    this.isConnected = false;
    this.reconnectAttempts = this.maxReconnectAttempts; // Prevent reconnect
  }

  /**
   * Check if the stream is currently connected.
   */
  get connected(): boolean {
    return this.isConnected;
  }
}

File: packages/python-sdk/src/mogbank/stream.py
"""Stream module — real-time balance and event streaming via WebSocket and SSE."""

from __future__ import annotations
import asyncio
import json
import logging
from typing import Optional, Callable, Awaitable, Any, Dict

import httpx

from mogbank.types import MogBankConfig

logger = logging.getLogger("mogbank.stream")


class StreamClient:
    """
    Real-time streaming client for MogBank.

    Supports:
    - WebSocket for balance updates, transaction events
    - SSE (Server-Sent Events) as fallback
    - Auto-reconnection with exponential backoff
    """

    def __init__(
        self,
        config: MogBankConfig,
        agent_id: Optional[str] = None,
        wallet_id: Optional[str] = None,
    ):
        self.config = config
        self.agent_id = agent_id or config.agent_id
        self.wallet_id = wallet_id

        ws_url = config.base_url.replace("https://", "wss://").replace("http://", "ws://")
        self.ws_url = f"{ws_url}/ws"
        self.sse_url = f"{config.base_url}/api/v1/stream"

        self._ws: Optional[Any] = None
        self._should_reconnect = True
        self._reconnect_delay = 1.0
        self._max_reconnect_delay = 60.0

        # Event handlers
        self._on_balance: Optional[Callable[[Dict[str, Any]], Awaitable[None]]] = None
        self._on_transaction: Optional[Callable[[Dict[str, Any]], Awaitable[None]]] = None
        self._on_error: Optional[Callable[[Exception], Awaitable[None]]] = None

    def on_balance(self, handler: Callable[[Dict[str, Any]], Awaitable[None]]):
        """Register a handler for real-time balance updates."""
        self._on_balance = handler

    def on_transaction(self, handler: Callable[[Dict[str, Any]], Awaitable[None]]):
        """Register a handler for transaction events."""
        self._on_transaction = handler

    def on_error(self, handler: Callable[[Exception], Awaitable[None]]):
        """Register a handler for stream errors."""
        self._on_error = handler

    async def connect_websocket(self) -> None:
        """Connect via WebSocket for real-time streaming."""
        try:
            import websockets
        except ImportError:
            logger.warning("websockets not installed, falling back to SSE")
            await self.connect_sse()
            return

        self._should_reconnect = True
        self._reconnect_delay = 1.0

        while self._should_reconnect:
            try:
                # Build connection URL with auth params
                url = self.ws_url
                params = []
                if self.agent_id:
                    params.append(f"agent_id={self.agent_id}")
                if self.wallet_id:
                    params.append(f"wallet_id={self.wallet_id}")
                if params:
                    url = f"{url}?{'&'.join(params)}"

                async with websockets.connect(url) as ws:
                    self._ws = ws
                    self._reconnect_delay = 1.0  # Reset on successful connection
                    logger.info(f"WebSocket connected: {url}")

                    async for message in ws:
                        try:
                            data = json.loads(message)
                            await self._handle_message(data)
                        except json.JSONDecodeError:
                            logger.warning(f"Invalid JSON from WebSocket: {message}")

            except asyncio.CancelledError:
                self._should_reconnect = False
                break
            except Exception as e:
                logger.warning(f"WebSocket disconnected: {e}")
                if self._on_error:
                    await self._on_error(e)

                if self._should_reconnect:
                    await asyncio.sleep(self._reconnect_delay)
                    self._reconnect_delay = min(
                        self._reconnect_delay * 2, self._max_reconnect_delay
                    )

    async def connect_sse(self) -> None:
        """Connect via SSE (Server-Sent Events) as a fallback."""
        self._should_reconnect = True
        self._reconnect_delay = 1.0

        params = []
        if self.agent_id:
            params.append(f"agent_id={self.agent_id}")
        if self.wallet_id:
            params.append(f"wallet_id={self.wallet_id}")
        url = self.sse_url
        if params:
            url = f"{url}?{'&'.join(params)}"

        while self._should_reconnect:
            try:
                async with httpx.AsyncClient(
                    timeout=httpx.Timeout(None),
                    headers={
                        "Accept": "text/event-stream",
                        "Cache-Control": "no-cache",
                    },
                ) as client:
                    async with client.stream("GET", url) as response:
                        response.raise_for_status()
                        self._reconnect_delay = 1.0
                        logger.info(f"SSE connected: {url}")

                        async for line in response.aiter_lines():
                            if not line.strip():
                                continue
                            if line.startswith("data:"):
                                data_str = line[len("data:"):].strip()
                                try:
                                    data = json.loads(data_str)
                                    await self._handle_message(data)
                                except json.JSONDecodeError:
                                    logger.warning(f"Invalid JSON from SSE: {data_str}")

            except Exception as e:
                logger.warning(f"SSE disconnected: {e}")
                if self._on_error:
                    await self._on_error(e)

                if self._should_reconnect:
                    await asyncio.sleep(self._reconnect_delay)
                    self._reconnect_delay = min(
                        self._reconnect_delay * 2, self._max_reconnect_delay
                    )

    async def _handle_message(self, data: Dict[str, Any]) -> None:
        """Dispatch incoming messages to registered handlers."""
        event_type = data.get("type", "")

        if event_type == "balance" and self._on_balance:
            await self._on_balance(data)
        elif event_type == "transaction" and self._on_transaction:
            await self._on_transaction(data)
        else:
            logger.debug(f"Unhandled event type: {event_type}")

    async def disconnect(self) -> None:
        """Disconnect the stream."""
        self._should_reconnect = False
        if self._ws:
            await self._ws.close()
            self._ws = None

    async def stream_balance(
        self, wallet_id: str
    ) -> None:
        """Begin streaming balance updates for a specific wallet."""
        self.wallet_id = wallet_id
        await self.connect_websocket()

    async def subscribe_transactions(self) -> None:
        """Subscribe to transaction events for the configured agent."""
        await self.connect_websocket()